lib/Backend/Vault/Kv/Internal.hs

```diff
@@ -0,0 +1,342 @@
+{-# LANGUAGE
+  TemplateHaskell
+, MultiParamTypeClasses
+, FlexibleInstances
+#-}
+
+module Backend.Vault.Kv.Internal
+  ( KvResponse (..)
+  , requestId, leaseId, renewable, leaseDuration, wrapInfo, warnings, auth
+  , ReadSecret (..)
+  , secret, createdTime, deletionTime, destroyed, version
+  , ListSecrets (..)
+  , PostSecret (..)
+  , cas
+  , PatchSecret (..)
+  , UpdateMetadata (..)
+  , maxVersions, casRequired, deleteVersionAfter
+  , VaultToken (..)
+
+  , ddata, customMetadata
+
+  , readSecret
+  , listSecrets
+  , postSecret
+  , patchSecret
+  , updateMetadata
+  , deleteSecret
+
+  , routes
+  )
+where
+
+import qualified Data.Text              as T;
+import qualified Data.HashMap.Strict    as HS;
+import qualified Data.Aeson             as A
+import qualified Data.Aeson.Types       as AT
+
+import           Data.Aeson             ((.:)) -- Control.Lens shadows (.=), use (A..=) instead
+import           Control.Applicative    ((<|>))
+import           Data.Proxy             (Proxy (Proxy))
+import           Servant.Client.Generic (AsClientT, genericClientHoist, genericClient)
+import           Control.Exception      (throwIO)
+
+import           Control.Lens
+import           Servant.API
+import           Servant.Client
+import           Servant.API.Generic
+
+-- |
+-- A type defining most responses from the Vault server, it wraps the data we're
+-- actually trying to read as some sort of super class.
+-- For example:
+--
+-- { "request_id": "00000000-0000-0000-0000-000000000000"
+-- , "lease_id": ""
+-- , "renewable": false
+-- , "lease_duration": 0
+-- , "data": <DATA>
+-- , "wrap_info": null
+-- , "warnings": null
+-- , "auth": null
+-- }
+data KvResponse a =
+  KvResponse
+  { _requestId :: T.Text
+  , _leaseId :: T.Text
+  , _renewable :: Bool
+  , _leaseDuration :: Int
+  , _kdata :: a
+  , _wrapInfo :: Maybe ()
+  , _warnings :: Maybe ()
+  , _auth :: Maybe ()
```

Comment:
Those three are guaranteed to be null? What if they suddenly aren't?

Reply:
They are not, but I couldn't figure out what they are supposed to be, or even could be as Vaults documentation sucks, much of this is reverse engineered... If someone manages to get a response from Vault where those 3 aren't null we can tackle it but right now all responses I've seen had those set to `null`